### ISSUE TITLE
On Locking and TemplateLocking

### ISSUE THREAD

**Issue by mtias** [WordPress/gutenberg#29864]
### Background

The block editor has had template locking support since very early on. Template locking allowed to designate the original content templates (the ones you could use on custom post types) to prevent insertion or moving blocks. Inner blocks also has supported a `templateLock` attribute. This was extended in #26128 to be used directly in columns and group as attributes, not just as internal props of `<InnerBlocks>`.

### Limitations

While powerful on its own, template locking has not offered too much granular control nor has the UI adapted to communicate different locking states.

With block themes, the necessity for locking down elements of the page is more pronounced. For example, it might be desirable that by default the "post-content" block of a `single` template is locked so that it cannot be removed. This extends to many other blocks and template parts in all sorts of ways.

### Proposal

Let's establish a new block-level API representing `lock` status. This is in addition to `templateLock`. The lock attribute would allow control over what UI interactions are present.

Lock status needs to allow a few state options. A locked block could imply any or all of these:
- Cannot be removed.
- Cannot be moved.
- Cannot be edited.
  - Worthwhile exploring (though complex) how specific attributes could be locked instead of the entire block. It might be better to keep this separate as a `lockAttribute` prop down the line.

While `templateLock` separately governs the inner blocks state (blocks cannot be added, etc).

This lock status can be toggled by the user. A user can choose to lock down a block so that they don't accidentally remove it, like on most design applications.

This could be represented in the UI as follows:

![image](https://user-images.githubusercontent.com/548849/111138818-abd36c80-8580-11eb-9955-74c6e5d167f0.png)

_Note: block locking has nothing to do with capabilities for editing and it only governs interface controls._

If a block lock state is editable by the user, clicking on the "lock" icon should bring about a modal with the lock configuration. If only "remove" is locked, the movers and drag handle should remain and the lock indication should go somewhere else (in the ellipsis menu, for example.)

Lock status should be reflected in more places than just the Toolbar:

- List View.
- Block Inspector.

![image](https://user-images.githubusercontent.com/548849/111139473-75e2b800-8581-11eb-8fd1-43ab980cb7a9.png)

**Comment by jameskoster** [comment]
Making templates contextually un-removable and/or un-editable (attributes) will be crucial in establishing an intuitive experience as one moves between editing content / template 👍 

Some more background on that (and rough concepts) can be found in https://github.com/WordPress/gutenberg/issues/27848.

> Cannot be moved.

This one feels especially tricky because the position of a block is always relative to those around it. Even if a block is position-locked to the top of the tree, it can still effectively be "moved" by _re-positioning the blocks around it_. Is this perhaps something that can only be applied to container blocks, and cascades down to all the innerblocks? IE everything is position-locked, or nothing is.

**Comment by mtias** [comment]
TemplateLocking already handles the disabling of moving for all children. I don't think it's much of an issue that you can move something by moving the things around it, that will always be the case. This is more about the peace of mind and reduced overhead of controls that you don't need.

**Comment by paaljoachim** [comment]
I really like the iterations you made Matias!
Simple and straight forward. 

I am wondering if it should contain a drop down of toggle options instead of a modal.
As one clicks the lock/unlock icon. A drop down is seen to where the administrator toggles various options on and off depending on what should be locked. Such as various levels of importance. Such as editing the block, moving, removing etc.

**Comment by nerrad** [comment]
> If a block lock state is editable by the user, clicking on the "lock" icon should bring about a modal with the lock configuration.

How would permissions governing editing block lock state be governed? It'd be nice if the API allowed for blocks to define the rules around that (either via user capability or programatically).

**Comment by senadir** [comment]
> TemplateLocking already handles the disabling of moving for all children. I don't think it's much of an issue that you can move something by moving the things around it, that will always be the case. This is more about the peace of mind and reduced overhead of controls that you don't need.

One good case I see for disabling moving is that if you lock 2+ blocks, they're always relatively on the same position, in our example (Checkout block), we always need the contact information section to be before address section, or the address section to be before the shipping options, locking them would prevent people from switching and creating a broken UX but still allow them to insert stuff between.

**Comment by senadir** [comment]
@mtias I started working on this yesterday, I'm going to tackle this by parts instead of all at once in a single PR.

Currently, the API lives on the `supports` level, with this shape
```
supports: {
  lock: {
    remove: true | false | 'user',
    move: true | false | 'user',
    edit: true | false | 'user',
  }
}
```

The values, if not set, would inherit the parent `templateLock` attribute.

One issue I'm not sure about is editing and how we intend to prevent this, it seems to be something on the block implementation side of things, not on block editor side. Moving and removing are done to the whole block at the data store level. Editing is done on parts of the block at the block level, best you can do is prevent `setAttributes` update from propagating or wrapping the whole block in a `Disabled` so users can't interact with it, but this would still leave settings accessible. 
All what we can do here is indicate to the block that the user has chosen to lock this block and the block has to honor its value.

The second question is where should this live, should it be on the Block type level (like the above example) or the instance level (block level)?

I can see the see that having this on the block type level would be the equivalence of `locked by system` in your example, while locking on the block level would be `locked by user` type of thing.

**Comment by jameskoster** [comment]
Sharing a couple of issues (with designs) here that may be relevant to the `edit` implementation.

In https://github.com/WordPress/gutenberg/pull/31461 the contents of blocks like the Post Title should be locked when editing a template.

Similarly, in https://github.com/WordPress/gutenberg/issues/32317 the contents of blocks like the Post Title should be locked when inside a Query block.

At a high level the edit-ability of the contents of these blocks boils down to whether the scope is local or not. If the source is outside of the immediate editing context then the block contents should be locked. I don't know if that can be built in to the API but I figured it was worth sharing.

**Comment by senadir** [comment]
This PR is an iteration of this issue
https://github.com/WordPress/gutenberg/pull/32457

**Comment by jorgefilipecosta** [comment]
> One issue I'm not sure about is editing and how we intend to prevent this, it seems to be something on the block implementation side of things, not on block editor side. 


If we could have edit locking as a general thing instead of a block-specific implementation it would have some advantages. For example, post-content may have multiple blocks inside and have an edit locking the edit locking should affect all the blocks inside even if they don't implement an edit lock. A general solution may be having the editor put an invisible overlay to make things unclickable and have something catch the focus so changing with the keyboard is also not an option.

**Comment by senadir** [comment]
>  A general solution may be having the editor put an invisible overlay to make things unclickable and have something catch the focus so changing with the keyboard is also not an option.

We use `<Disabled>` to prevent interactions but this still doesn't prevent settings from showing up in the sidebar, do you want to prevent both?

**Comment by jorgefilipecosta** [comment]
> We use <Disabled> to prevent interactions but this still doesn't prevent settings from showing up in the sidebar, do you want to prevent both?

I guess if we add a general API to the core the slot fills of the sidebar and toolbar would not render if the block is disabled.

But at the start, we may opt for a block-specific implementation as you referred using disabled and a condition to not render sidebar and toolbar. This solution is simple and also covers the child block case because with <Disabled> one can not select the child blocks.

**Comment by paaljoachim** [comment]
Is there anything standing in the way or is needed to get done before we can move onward with the 
**Add lock feature for Reusable Blocks to protect inner blocks** PR?
https://github.com/WordPress/gutenberg/pull/32710

**Comment by senadir** [comment]
Following up on https://github.com/WordPress/gutenberg/pull/32457, this is the list of items remaining to achieve full locking:
- [ ] Update Lock API to support user controlled locking.
- [ ] Add a modal to toggle locking for moving and removing.
- [ ] Show Lock icon on block list view and toolbar.
- [ ] Remove `core/lock/addAttribute` filter - #35389
- [ ] Disable moving in block list view.

**Comment by Mamaduka** [comment]
I am super excited about this feature.

@senadir, let me know if you need help with any remaining items.

**Comment by senadir** [comment]
@Mamaduka feel free to pick up from the list as I'm struggling to prioritize GB contributions outside my core job and might take some time before I fully work on them.

**Comment by Mamaduka** [comment]
I'm starting to work on the remaining items for this feature.

One thing that isn't very clear for me is the definitions of "Locked (by user)" and "Locked (by the system)." I want to get this right so that I can iterate faster.

Below is my attempt to formalize these definitions:

* "Locked (by the system)" - when `lock` attribute values are set to `true` on a block level.
* "Locked (by user)" - when the `lock` attribute isn't set, it can't be changed by a user via UI. Possible values: `"user"|false`.

Is this definition correct?

cc @mtias, @senadir

**Comment by senadir** [comment]
I think we should first agree on something, do we want to enable user locking on all blocks (opt out) or should it be an opt in thing (an extra attribute/different value defined on the block instance or block type).

### Opt out
1. All blocks can be locked, via an option in toolbar maybe, that would set the block `lock.move` or `lock.remove` to true or false. This is user locking.
2. Locking on a system level would be that locking is defined on the attribute level `lock.move` and there's maybe another attribute `allowUserLocking` or `support.userLocking` that would show/prevent locking behavior.

### Opt in
- Just point 2 from above, but you can't make a regular block user locked, it has be first allowed on the attributes level or block type level.

**Comment by carlomanf** [comment]
> With block themes, the necessity for locking down elements of the page is more pronounced. For example, it might be desirable that by default the "post-content" block of a `single` template is locked so that it cannot be removed. This extends to many other blocks and template parts in all sorts of ways.

This part concerns me because it seems to again blur the line between plugins and themes. If block themes have the ability to use the lock attribute to "protect the user from themselves" when customising templates, then I can see it turning into a confusing mess where some block themes use them liberally and others don't use them at all.

Editing capability should be governed by plugins and core, not on a theme-by-theme basis. This is better for everyone, because site administrators don't face the frustration of being treated like novice users by "well-meaning" themes, and they also have the ability to switch themes freely without it dismantling the existing restrictions they have set up for their users.

What plans are there to address this concern?

**Comment by mtias** [comment]
@Mamaduka sorry for the delay responding, finally able to spend some time on this :)

- **L. by system**: this means the block is locked but the user doesn't have permissions to change that state. These permissions could come from user roles & capabilities or by flags in theme.json (which could also be based on roles or capabilities).
- **L. by user**: just means that it's self-service, where you can lock stuff for convenience like on design tools. Clicking on the lock icon would show a menu or modal where you can set the different properties of how to lock / unlock that block.

The second case can be useful for users dealing with reusable blocks, for example.

@carlomanf that's a very valid concern. I'd expect places like the theme directory or patterns directory would have guidelines to not lock things down at all. The availability of mechanisms to restrict things becomes more relevant for custom themes (that is, specific sites that are maintained and managed in specific ways), even if they are present in theme.json. The same is true for appearance tools, where a theme being super opinionated is fine in the context of specific sites that need some branding boundaries, but is not good for a multi-purpose theme shared broadly.

**Comment by Mamaduka** [comment]
Thank you, @mtias.

> These permissions could come from user roles & capabilities or by flags in theme.json (which could also be based on roles or capabilities).

Do you have an existing capability or role in mind for this, or should we introduce a new capability?

**Comment by critterverse** [comment]
>If a block lock state is editable by the user, clicking on the "lock" icon should bring about a modal with the lock configuration.

I took a pass on the modal mentioned above and how it might work in the lock/unlock flow.  Here’s how the design of the modal could look:

<img width="100%" alt="modal" src="https://user-images.githubusercontent.com/77359364/155371335-f13c7b20-d1a0-4ffa-bca0-1a151aa4b9eb.png">

- It uses a checklist system like the one used in Preferences → Blocks → Visible Blocks
- We could use the block type in the title of the modal

Here’s the proposed flow:

![Reusable block — locking](https://user-images.githubusercontent.com/77359364/155375201-ff8c13da-1b1e-4f02-ad63-86c31037912f.png)

1. Start editing a block
2. Find the option to lock within the ellipses menu in the block toolbar
3. A modal appears with a description and options for locking; all options are selected by default
4. The user can make their selections and “Apply” the changes
5. When the settings have been applied from the modal, a lock icon will appear in the block toolbar (movers and drag handle are removed)
6. Selecting the lock icon in the toolbar will re-open the same modal with the user's preferences selected
7. When the block is locked, the available action in the ellipses menu is “Unlock”
8. When unlocked, the lock icon disappears from the toolbar (drag handles and movers reappear)

[Figma file](https://www.figma.com/file/CUZSVCxP0sDGlP8aInzGaO/User-level-Block-Locking?node-id=20%3A2763)

**Comment by leclaeli** [comment]
I just want to echo my support for more granular locking control. I work at a large university with several hundred WP sites under one multisite network. We have hundreds of content authors so being able to control what can be edited is an ongoing concern.

As we move sites over to a more block based theme we have run into issues with reusable blocks being unintentionally updated and have mostly abandoned using them because of this :(. The newly added lock attribute and this proposal will really help out!

**Comment by Mamaduka** [comment]
Thank you, @critterverse, for the fantastic design update 🤩 

I plan to start working on the UI parts of this feature next week.

**Comment by carlomanf** [comment]
> * **L. by user**: just means that it's self-service, where you can lock stuff for convenience like on design tools. Clicking on the lock icon would show a menu or modal where you can set the different properties of how to lock / unlock that block.
> 
> The second case can be useful for users dealing with reusable blocks, for example.

I think there needs to be a distinction between unlocking in the browser, and removing the lock persistently. If a user unlocks a reusable block to edit it, then saves and closes the editor, it should revert back to locked again the next time they come back to edit. Without this distinction, the onus is on the user to remember to re-lock when they are done. The mock-up by @critterverse seems to be lacking this nuance as well, particularly screen number 6, which I expected would unlock the block in the browser, but it seems to bring up the option to remove the lock persistently.

**Comment by critterverse** [comment]
>I think there needs to be a distinction between unlocking in the browser, and removing the lock persistently. If a user unlocks a reusable block to edit it, then saves and closes the editor, it should revert back to locked again the next time they come back to edit.

Interesting, @carlomanf. It seems like the expected behavior around this may vary (personally, I would expect the editor to remember my most recently saved lock state) but I'm sure there are other possibilities to be explored that could help with remembering to re-lock. For example, maybe the multi-entity saving panel could eventually contain some indication of the lock status.

My vote would be to keep the UI implementaton as simple as possible for the first iterations of this but I can imagine there being lots of opportunities to keep building on the functionality from there.

**Comment by jameskoster** [comment]
> I would expect the editor to remember my most recently saved lock state

I wonder how the lock state is saved. If it's attached to the document (rather than the user) this might be tricky.

I do agree that the expected behavior can be subjective though. Perhaps another option would be to have a setting or a filter for the default state of certain block types?

**Comment by Mamaduka** [comment]
> I wonder how the lock state is saved.

It's attached to a document and saved as [block attributes](https://make.wordpress.org/core/2022/01/08/locking-blocks-in-wordpress-5-9/).

**Comment by pablohoneyhoney** [comment]
Nice! 

- Do you think we could try to have less separators in the modal?
- Have we considered calling it _Restrict (block name)_? So it refers to adding granular restrictions rather than only _locking_ it as a whole, which is what it reads when seen in the dropdown. 
- Perhaps a bit more line height in the modal text. 
- And explore a more discreet _restricted_ icon? While is clear it looks a bit jarring. Perhaps it's a smaller version of the current icon that makes sense in other contexts but it might be a bit too large for these. 

<img width="336" alt="Screen Shot 2022-03-03 at 3 51 08 PM" src="https://user-images.githubusercontent.com/25516611/156650854-1ca3a25e-1085-4348-8631-e144eada721b.png">

**Comment by pablohoneyhoney** [comment]
As an aside, the icons seem to difer visually, not sure why. The below one (still large I think) seems more proportionally right, less a _suitcase_. 

<img width="344" alt="Screen Shot 2022-03-03 at 3 54 57 PM" src="https://user-images.githubusercontent.com/25516611/156651463-9874a18f-6b18-425e-bb46-866b26c1a9fd.png">

**Comment by critterverse** [comment]
Thanks for the notes @pablohoneyhoney — we'll work on refining some of the design aspects in https://github.com/WordPress/gutenberg/pull/39183.

>Have we considered calling it Restrict (block name)? So it refers to adding granular restrictions rather than only locking it as a whole, which is what it reads when seen in the dropdown.

This does probably make more sense 😆 Folks seem to be understanding the word "lock" pretty well in this context so far, too. What do others think?

**Comment by jasmussen** [comment]
Solid feedback. @Mamaduka I'm happy to help you with these items next week. I'll also update the icons in Figma where they are both chunky:
<img width="376" alt="Screenshot 2022-03-04 at 07 18 50" src="https://user-images.githubusercontent.com/1204802/156710584-e90e5c29-9900-4c7c-86e0-97752d30c846.png">

**Comment by jameskoster** [comment]
There are a few unpublished icon explorations in the [WordPress Design Library](https://www.figma.com/file/e4tLacmlPuZV47l7901FEs/WordPress-Design-Library?node-id=28932%3A23409).

<img width="400" alt="Screenshot 2022-03-04 at 11 57 29" src="https://user-images.githubusercontent.com/846565/156759603-c5394b9e-88c9-4ff5-b4df-fc13ca8bb561.png">

**Comment by jasmussen** [comment]
I've made a PR to update the icons: https://github.com/WordPress/gutenberg/pull/39252

**Comment by gaambo** [comment]
> L. by system: this means the block is locked but the user doesn't have permissions to change that state. These permissions could come from user roles & capabilities or by flags in theme.json (which could also be based on roles or capabilities).

For building custom themes this is very important. Right now we're using patterns with lock-attributes + templateLock-attributes to allow the editors to insert specific rows/sections which were designed in this specific way. So in this case, the lock-UI should not be visible. This could also be handled by a specific capability (probably to complicated). But I think there needs to be some way to lock use the block-locking in a programmatic way only and disable the lock-UI. 
I'm wondering how granular this control should be. If there's a flag in theme.json it disables the lock-UI for alle blocks. Maybe have the `lock` attribute have another property `allow_change` or something? So each pattern/template which uses block-locking can define if it's a "system" lock or not.

**Comment by Mamaduka** [comment]
## Locked by System

The easiest way to implement this restriction is to use block editor settings. It will be enabled by default but will give developers the flexibility to toggle the feature as they see fit.

We can also support passing value from the theme.json.

## Examples

```php
add_filter(
	'block_editor_settings_all',
	static function( $settings, $context ) {
		// Allow for the Editor role and above.
		$settings['__experimentalAllowBlockLock'] = current_user_can( 'delete_others_posts' );

		// Disable for Pages.
		if ( $context->post && $context->post->post_type === 'page' ) {
			$settings['__experimentalAllowBlockLock'] = false;
		}
		
		// Only enable for specific user(s).
		$user = wp_get_current_user();
		$settings['__experimentalAllowBlockLock'] = $user->ID === 1;


		return $settings;
	},
	10,
	2
);
```

**Comment by senadir** [comment]
This implementation of locked by system seems to reflect the ability to access locking **on all blocks**. This doesn't offer any options at all, you can either lock/unlock all blocks or nothing.

Shouldn't system locking be a bit granular than this? on a block type level?

**Comment by Mamaduka** [comment]
@senadir I think we can also include your original suggestion about block level settings.

> Locking on a system level would be that locking is defined on the attribute level lock.move and there's maybe another attribute allowUserLocking or support.userLocking that would show/prevent locking behavior.

@mtias what do you think?

**Comment by fabiankaegy** [comment]
@Mamaduka I like the idea with presenting the global setting that way. Especially the Idea to do it via the php filter because often times it really needs to be dependent on a user role or on a specific user id. Which is not something that would be possible in `theme.json` alone. 

And yeah allowing this setting again on a per block basis in the `block_editor_settings` would also be quite interesting. That way you could open up some blocks but leave others locked. Which I can see being very useful.

**Comment by Mamaduka** [comment]
> And yeah allowing this setting again on a per block basis in the block_editor_settings would also be quite interesting. That way you could open up some blocks but leave others locked. Which I can see being very useful.

Using block supports for this makes more sense to me. Blocks can define if locking UI should be available for them like @senadir suggested.

**Comment by Mamaduka** [comment]
I create a new PR for system lock settings - #39566.

**Comment by Mamaduka** [comment]
@critterverse, I plan to continue work started in #32710 and add edit locking support for Reusable Blocks.

Should blocks have any other indicators that editing is disabled besides the lock icon? I'm asking this because when UI is disabled, it might be confusing why the user can't edit block.

**Comment by critterverse** [comment]
>I plan to continue work started in https://github.com/WordPress/gutenberg/pull/32710 and add edit locking support for Reusable Blocks.

Nice! For user locking, we could include the “restrict editing” option in the modal for Reusable blocks:

<img alt="Locking" src="https://user-images.githubusercontent.com/77359364/160022072-81a723ce-f2fd-42e9-acc2-7b94bee46ca0.png">

Once locked, are we thinking it would work like https://github.com/WordPress/gutenberg/pull/32710#issuecomment-862247159 where you can’t select the inner blocks?

It seems like the clickthrough overlay could be really helpful here. Just a thought but I wonder if it could work slightly differently when editing is locked. For example, rather than an overlay effect on hover and the first click "unlocking" access to the inner contents, maybe the hover overlay (with no clickthrough) could remain even after selecting the Reusable block.

**Comment by Mamaduka** [comment]
That's an interesting idea. So I will start with that, and let's see how it works in practice.

**Comment by critterverse** [comment]
@Mamaduka It could be nice to start minimal and see if the clickthrough or anything else is needed from there!

**Comment by Mamaduka** [comment]
Hi, folks

I started working on the edit Locking ability for the Reusable Blocks. You can follow the progress in this PR - #39950.

Cc @syedbalkhi

**Comment by Mamaduka** [comment]
A summary based on the Block edit locking discussion from the [#core-editor Slack channel (needs login)](https://wordpress.slack.com/archives/C02QB2JS7/p1650367388054399):

* The general feeling is that the feature should be available for all blocks, not restricted to Reusable.
* We need a generic method to disable editing. For example, the content overlay treatment might not work for every block.
* The disabled content must be equally accessible as any other content on the page.

With this in mind, and considering issues we're facing with [content overlay](https://github.com/WordPress/gutenberg/pull/40339), I want to propose removing Reusable block edit locking from WP 6.0 release. Continue the feature development in the plugin and ship it with the next release.

**Comment by adamziel** [comment]
@Mamaduka I agree, let's punt block editing locking to the next release. Would you please prepare a PR for that?

**Comment by Mamaduka** [comment]
@mtias, can we close this issue?

**Comment by MadtownLems** [comment]
I'd like to help make a case for the ability prevent editing of any block, whether reusable or not.

We run networks of sites for educators who aren't "web designers", but still have Site Administrator levels of access on their sites. We'd love for the ability to easily lock a section of their page content so they can't break it - even accidentally!

Another potential use of this is MultiSite where only Network Administrators have access to unfiltered_html. If we could lock block CONTENT, we'd be able to do things like easily put iframes or custom scripts onto a page via Custom HTML block, then lock it. This could allow the Site Administrator the ability to manage all the content AROUND our Custom HTML block without breaking it when they save the page (and scripts/iframes get stripped).

**Comment by mtias** [comment]
@MadtownLems that should be possible already for any container block. Isn't that the case for you or is the use case a bit different?

**Comment by MadtownLems** [comment]
Was this functionality added in 6.2? If so, I haven't had a chance to test yet. My comment was from 6.1

**Comment by mtias** [comment]
Locking on any block has been there for a while, but maybe I misunderstood the question — did you mean for locking to also prevent content editing here? 

<img width="679" alt="image" src="https://github.com/WordPress/gutenberg/assets/548849/cf15297b-34da-48ae-9d72-1d03247477c8">

**Comment by MadtownLems** [comment]
> did you mean for locking to also prevent content editing here?

Yes, exactly.  To clarify the use case:
MultiSite, where only Network Administrators can do things like use iframes.  If a Network Administrator puts an iframe on a page, and then a normal Site Administrator edits it, the iframe will get stripped out.

We'd love to be able to have Network Administrators add things like a Custom HTML block to a page (that includes an iframe, script, etc) - then Lock the content. In an ideal world, a Site Administrator wouldn't be able to unlock it (that part is easy enough), but would still be able to make edits to the rest of the page without worrying about the locked content being modified or stripped away in any way.

**Comment by mtias** [comment]
Got it, I think a full content lock would be a nice and easy addition to the API. Would you mind opening a new issue describing your use case?

**Comment by Zilero232** [comment]
How can I block a custom block so that it cannot be added to the group?
